File: cli/internal/commands/books/management.go
// Package books — phase 4B: the management layer.
//
// `analytique` and `tax` are pure reads over derivations the server computes
// at request time. `analyse record` is the agent write-back contract: an
// outside agent reads the data, answers a question, and files the answer WITH
// its based_on snapshot. The record is permanent — there is no analyse edit
// and no analyse delete, on purpose, and the server enforces it by grant.
// `category create` is configuration: the breakdown's buckets, per book.
package books

import (
	"encoding/json"
	"fmt"
	"io"
	"strconv"

	"github.com/blackcode-switzerland/bc-issues/cli/internal/client"
	"github.com/blackcode-switzerland/bc-issues/cli/internal/cmdutil"
	"github.com/blackcode-switzerland/bc-issues/cli/internal/output"
	"github.com/spf13/cobra"
)

// biText renders a bilingual {fr, en} value (or a plain string) for a card.
// English first when both halves exist — the CLI's own copy is English.
func biText(v any) string {
	switch t := v.(type) {
	case string:
		return t
	case map[string]any:
		if en, ok := t["en"].(string); ok && en != "" {
			return en
		}
		if fr, ok := t["fr"].(string); ok && fr != "" {
			return fr
		}
	}
	return ""
}

// ---------------------------------------------------------------------------
// bk books analytique
// ---------------------------------------------------------------------------

func newAnalytiqueCmd() *cobra.Command {
	var scope client.BooksScope
	cmd := &cobra.Command{
		Use:         "analytique",
		Annotations: map[string]string{"routes": "GET /api/workspaces/{ws}/analytique"},
		Short:       "Cost breakdown per category, and the monthly flows",
		Long: "The management view of one (book, exercice): charges per category and the\n" +
			"monthly produits/charges series. Derived from POSTED lines at request time —\n" +
			"staged money reaches no chart. A simplified book groups its dépenses by their\n" +
			"own category label instead.",
		Args: cobra.NoArgs,
		RunE: func(cmd *cobra.Command, args []string) error {
			format, err := output.Resolve(cmd)
			if err != nil {
				return err
			}
			c, ws, err := clientAndWorkspace()
			if err != nil {
				return err
			}
			r, err := c.GetBooksAnalytique(ws, scope)
			if err != nil {
				return err
			}
			return output.Render(format, r, func(w io.Writer) error {
				fmt.Fprintf(w, "%s · %d\n\n", r.Entity, r.Exercice)
				tw := output.Tabwriter(w)
				fmt.Fprintln(tw, "CATEGORY\tLABEL\tAMOUNT\tLINES")
				for _, cat := range r.Categories {
					fmt.Fprintf(tw, "%s\t%s\t%s\t%d\n", cat.Key, cmdutil.Truncate(biText(cat.Label), 28), cat.Amount, len(cat.Lines))
				}
				if err := tw.Flush(); err != nil {
					return err
				}
				fmt.Fprintln(w)
				tw = output.Tabwriter(w)
				fmt.Fprintln(tw, "MONTH\tPRODUITS\tCHARGES")
				for _, f := range r.MonthlyFlows {
					fmt.Fprintf(tw, "%s\t%s\t%s\n", f.Month, f.Produits, f.Charges)
				}
				return tw.Flush()
			})
		},
	}
	scopeFlags(cmd, &scope)
	return cmd
}

// ---------------------------------------------------------------------------
// bk books analyse — list, show, record
// ---------------------------------------------------------------------------

func newAnalyseCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "analyse",
		Short: "Recorded analyses — questions asked, verdicts filed, snapshots kept",
	}
	cmd.AddCommand(newAnalyseListCmd(), newAnalyseShowCmd(), newAnalyseRecordCmd())
	return cmd
}

func newAnalyseListCmd() *cobra.Command {
	var entity string
	cmd := &cobra.Command{
		Use:         "list",
		Annotations: map[string]string{"routes": "GET /api/workspaces/{ws}/analyses"},
		Short:       "List filed analyses",
		Args:        cobra.NoArgs,
		RunE: func(cmd *cobra.Command, args []string) error {
			format, err := output.Resolve(cmd)
			if err != nil {
				return err
			}
			c, ws, err := clientAndWorkspace()
			if err != nil {
				return err
			}
			rows, err := c.ListBooksAnalyses(ws, entity)
			if err != nil {
				return err
			}
			return output.Render(format, rows, func(w io.Writer) error {
				tw := output.Tabwriter(w)
				fmt.Fprintln(tw, "#\tASKED\tBOOK\tBY\tAGENT\tQUESTION")
				for _, a := range rows {
					asked := a.Asked
					if len(asked) > 10 {
						asked = asked[:10]
					}
					fmt.Fprintf(tw, "%d\t%s\t%s\t%s\t%s\t%s\n",
						a.Number, asked, a.Entity, a.AskedBy, a.Agent, cmdutil.Truncate(biText(a.Question), 44))
				}
				if err := tw.Flush(); err != nil {
					return err
				}
				if len(rows) == 0 {
					fmt.Fprintln(cmd.ErrOrStderr(), "(no analyses filed)")
				}
				return nil
			})
		},
	}
	cmd.Flags().StringVar(&entity, "entity", "", "Only this book's analyses")
	return cmd
}

func newAnalyseShowCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:         "show <number>",
		Annotations: map[string]string{"routes": "GET /api/workspaces/{ws}/analyses/{number}"},
		Short:       "One analysis, whole: verdict, figures, and the based_on snapshot as filed",
		Args:        cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			format, err := output.Resolve(cmd)
			if err != nil {
				return err
			}
			n, err := strconv.Atoi(args[0])
			if err != nil || n < 1 {
				return fmt.Errorf("%q is not an analysis number; bk books analyse list shows them", args[0])
			}
			c, ws, err := clientAndWorkspace()
			if err != nil {
				return err
			}
			a, err := c.GetBooksAnalysis(ws, n)
			if err != nil {
				return err
			}
			return output.Render(format, a, func(w io.Writer) error {
				fmt.Fprintf(w, "Analysis #%d · %s · asked %s by %s · agent %s\n", a.Number, a.Entity, a.Asked, a.AskedBy, a.Agent)
				if s := biText(a.ScenarioLabel); s != "" {
					fmt.Fprintf(w, "Scenario: %s", s)
					if a.RunwayAfterMonths != nil {
						fmt.Fprintf(w, " (runway after: %.1f months)", *a.RunwayAfterMonths)
					}
					fmt.Fprintln(w)
				}
				fmt.Fprintf(w, "\nQ: %s\n", biText(a.Question))
				fmt.Fprintf(w, "A: %s\n", biText(a.Verdict))
				if len(a.Figures) > 0 {
					fmt.Fprintln(w, "\nFigures:")
					for _, f := range a.Figures {
						fmt.Fprintf(w, "  %s: %v\n", biText(f["label"]), f["value"])
					}
				}
				if len(a.BasedOn) > 0 {
					fmt.Fprintln(w, "\nBased on (snapshot at answer time, never recomputed):")
					for _, b := range a.BasedOn {
						line := fmt.Sprintf("  %s: %v", biText(b["label"]), b["value"])
						if href, ok := b["href"].(string); ok && href != "" {
							line += "  → " + href
						}
						fmt.Fprintln(w, line)
					}
				}
				return nil
			})
		},
	}
	return cmd
}

func newAnalyseRecordCmd() *cobra.Command {
	var req client.RecordBooksAnalysisRequest
	var question, verdict, scenario, figuresJSON, basedOnJSON string
	var runway float64
	cmd := &cobra.Command{
		Use:         "record",
		Annotations: map[string]string{"routes": "POST /api/workspaces/{ws}/analyses"},
		Short:       "File an analysis — the agent write-back",
		Long: "File one analysis: the question, the verdict, and the based_on snapshot of\n" +
			"what was READ to answer it. The row is permanent the moment it lands: there is\n" +
			"no edit and no delete. A drifted answer is re-asked into a NEW record, and both\n" +
			"stand.\n\n" +
			"--based-on and --figures take JSON arrays; every based_on item needs label and\n" +
			"value ({\"label\": ..., \"value\": ..., \"href\"?: ...}) — the server refuses a\n" +
			"snapshot that snapshots nothing.",
		Args: cobra.NoArgs,
		RunE: func(cmd *cobra.Command, args []string) error {
			format, err := output.Resolve(cmd)
			if err != nil {
				return err
			}
			c, ws, err := clientAndWorkspace()
			if err != nil {
				return err
			}
			req.Question = question
			req.Verdict = verdict
			if scenario != "" {
				req.ScenarioLabel = scenario
			}
			if cmd.Flags().Changed("runway-after-months") {
				req.RunwayAfterMonths = &runway
			}
			if figuresJSON != "" {
				if err := json.Unmarshal([]byte(figuresJSON), &req.Figures); err != nil {
					return fmt.Errorf("--figures is not a JSON array: %w", err)
				}
			}
			if basedOnJSON != "" {
				if err := json.Unmarshal([]byte(basedOnJSON), &req.BasedOn); err != nil {
					return fmt.Errorf("--based-on is not a JSON array: %w", err)
				}
			}
			a, err := c.RecordBooksAnalysis(ws, req)
			if err != nil {
				return err
			}
			return output.Render(format, a, func(w io.Writer) error {
				fmt.Fprintf(w, "analysis #%d filed on %s — permanent, with its snapshot\n", a.Number, a.Entity)
				return nil
			})
		},
	}
	cmd.Flags().StringVar(&req.Entity, "entity", "", "The book this analysis is about (required)")
	cmd.Flags().StringVar(&req.AskedBy, "asked-by", "", "Who asked the question (required)")
	cmd.Flags().StringVar(&req.Agent, "agent", "", "Which agent answered (required)")
	cmd.Flags().StringVar(&question, "question", "", "The question, as asked (required)")
	cmd.Flags().StringVar(&verdict, "verdict", "", "The verdict, as given (required)")
	cmd.Flags().StringVar(&figuresJSON, "figures", "", `Headline figures, JSON: [{"label": ..., "value": ...}]`)
	cmd.Flags().StringVar(&basedOnJSON, "based-on", "", `What was read, JSON: [{"label": ..., "value": ..., "href"?: ...}]`)
	cmd.Flags().StringVar(&scenario, "scenario", "", "Optional what-if label")
	cmd.Flags().Float64Var(&runway, "runway-after-months", 0, "Numeric runway restatement, for charts")
	_ = cmd.MarkFlagRequired("entity")
	_ = cmd.MarkFlagRequired("asked-by")
	_ = cmd.MarkFlagRequired("agent")
	_ = cmd.MarkFlagRequired("question")
	_ = cmd.MarkFlagRequired("verdict")
	return cmd
}

// ---------------------------------------------------------------------------
// bk books category — list, create
// ---------------------------------------------------------------------------

func newCategoryCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "category",
		Short: "The analytique's cost buckets — per book, account-mapped",
	}
	cmd.AddCommand(newCategoryListCmd(), newCategoryCreateCmd())
	return cmd
}

func newCategoryListCmd() *cobra.Command {
	var entity string
	cmd := &cobra.Command{
		Use:         "list",
		Annotations: map[string]string{"routes": "GET /api/workspaces/{ws}/analytique/categories"},
		Short:       "List a book's cost categories",
		Args:        cobra.NoArgs,
		RunE: func(cmd *cobra.Command, args []string) error {
			format, err := output.Resolve(cmd)
			if err != nil {
				return err
			}
			c, ws, err := clientAndWorkspace()
			if err != nil {
				return err
			}
			rows, err := c.ListBooksCategories(ws, entity)
			if err != nil {
				return err
			}
			return output.Render(format, rows, func(w io.Writer) error {
				tw := output.Tabwriter(w)
				fmt.Fprintln(tw, "#\tKEY\tLABEL\tACCOUNTS\tRETIRED")
				for _, r := range rows {
					retired := ""
					if r.Retired {
						retired = "retired"
					}
					fmt.Fprintf(tw, "%d\t%s\t%s\t%s\t%s\n",
						r.Number, r.Key, cmdutil.Truncate(biText(r.Label), 28), cmdutil.Truncate(fmt.Sprint(r.Accounts), 30), retired)
				}
				return tw.Flush()
			})
		},
	}
	cmd.Flags().StringVar(&entity, "entity", "", "Which book (defaults to the first)")
	return cmd
}

func newCategoryCreateCmd() *cobra.Command {
	var entity, key, labelFr, labelEn string
	var accounts []string
	cmd := &cobra.Command{
		Use:         "create",
		Annotations: map[string]string{"routes": "POST /api/workspaces/{ws}/analytique/categories"},
		Short:       "Create a cost category",
		Long: "Create one cost category: a key, a label, and the ledger accounts it counts.\n\n" +
			"Two refusals guard it: every account must exist in the book's chart, and no\n" +
			"account may already be counted by another active category — one franc, one bar.",
		Args: cobra.NoArgs,
		RunE: func(cmd *cobra.Command, args []string) error {
			format, err := output.Resolve(cmd)
			if err != nil {
				return err
			}
			c, ws, err := clientAndWorkspace()
			if err != nil {
				return err
			}
			var label any
			switch {
			case labelFr != "" && labelEn != "":
				label = map[string]any{"fr": labelFr, "en": labelEn}
			case labelFr != "":
				label = labelFr
			default:
				label = labelEn
			}
			r, err := c.CreateBooksCategory(ws, client.CreateBooksCategoryRequest{
				Entity: entity, Key: key, Label: label, Accounts: accounts,
			})
			if err != nil {
				return err
			}
			return output.Render(format, r, func(w io.Writer) error {
				fmt.Fprintf(w, "category #%d %q on %s, counting %v\n", r.Number, r.Key, r.Entity, r.Accounts)
				return nil
			})
		},
	}
	cmd.Flags().StringVar(&entity, "entity", "", "The book (required)")
	cmd.Flags().StringVar(&key, "key", "", "Stable key: lowercase, digits, underscores (required)")
	cmd.Flags().StringVar(&labelFr, "label-fr", "", "French label")
	cmd.Flags().StringVar(&labelEn, "label-en", "", "English label (at least one label required)")
	cmd.Flags().StringSliceVar(&accounts, "accounts", nil, "Ledger accounts this bucket counts, comma-separated (required)")
	_ = cmd.MarkFlagRequired("entity")
	_ = cmd.MarkFlagRequired("key")
	_ = cmd.MarkFlagRequired("accounts")
	return cmd
}

// ---------------------------------------------------------------------------
// bk books tax
// ---------------------------------------------------------------------------

func newTaxCmd() *cobra.Command {
	var scope client.BooksScope
	cmd := &cobra.Command{
		Use:         "tax",
		Annotations: map[string]string{"routes": "GET /api/workspaces/{ws}/tax-snapshot"},
		Short:       "The PM tax snapshot — derived, cited, never stored",
		Long: "VAT position from the entries' own TVA columns, profit and equity from the\n" +
			"statements, and the two PM tax ESTIMATES from the book's parameter record.\n" +
			"A book with no parameters answers 'not configured' — never someone else's\n" +
			"rates. Tracking over time is b/tax; this is a snapshot only.",
		Args: cobra.NoArgs,
		RunE: func(cmd *cobra.Command, args []string) error {
			format, err := output.Resolve(cmd)
			if err != nil {
				return err
			}
			c, ws, err := clientAndWorkspace()
			if err != nil {
				return err
			}
			r, err := c.GetBooksTaxSnapshot(ws, scope)
			if err != nil {
				return err
			}
			return output.Render(format, r, func(w io.Writer) error {
				fmt.Fprintf(w, "%s · %d\n\n", r.Entity, r.Exercice)
				fmt.Fprintf(w, "Résultat: %s   Capitaux propres: %s\n", r.Profit, r.Equity)
				if r.Vat != nil {
					fmt.Fprintf(w, "\nTVA — opening due %s, output YTD %s, input claimed YTD %s → net due %s\n",
						r.Vat.OpeningDue, r.Vat.OutputYtd, r.Vat.InputClaimedYtd, r.Vat.NetDue)
				} else {
					fmt.Fprintln(w, "\nTVA — not registered")
				}
				if r.Tax == nil {
					fmt.Fprintln(w, "\nTax parameters: NOT CONFIGURED for this book — no rates are shown rather than someone else's.")
					return nil
				}
				t := r.Tax
				fmt.Fprintf(w, "\nImpôt sur le bénéfice (%s / %s) — estimate:\n", t.Canton, t.Commune)
				fmt.Fprintf(w, "  cantonal %s + communal %s + IFD %s = %s\n", t.ProfitTax.Cantonal, t.ProfitTax.Communal, t.ProfitTax.Ifd, t.ProfitTax.Total)
				fmt.Fprintf(w, "  statutory %.2f%% · effective %.2f%% (taxes are deductible)\n", t.ProfitTax.StatutoryPct, t.ProfitTax.EffectivePct)
				fmt.Fprintf(w, "Impôt sur le capital — gross %s, credited against profit tax %s → net due %s\n",
					t.CapitalTax.Gross, t.CapitalTax.Credited, t.CapitalTax.NetDue)
				if cap, ok := t.Params["capital_tax"].(map[string]any); ok {
					if confirmed, ok := cap["confirmed"].(bool); ok && !confirmed {
						fmt.Fprintln(w, "  (capital tax UNCONFIRMED: the art. 118 imputation question is open with the fiduciary)")
					}
				}
				return nil
			})
		},
	}
	scopeFlags(cmd, &scope)
	return cmd
}

File: apps/books/lib/derive/management.ts
// Phase 4B derivations: monthly flows, the cost breakdown, the VAT position
// and the two PM tax estimates. NOTHING HERE IS EVER STORED — index.ts's rule.
//
// ===========================================================================
// PORTED FROM THE MOCKUP. READ ITS VERSION BEFORE CHANGING THIS ONE.
// ===========================================================================
// Reference: `bbooks/assets/bbooks-data.js` — monthlyFlows (l.392),
// costBreakdownFor (l.421), pmProfitTax (l.1272), pmCapitalTax (l.1289),
// vatPosition (l.1443) in the `b-mockups` repo.
//
// Money that comes from POSTINGS is summed in centimes, in bigint, exactly as
// index.ts does it: flows, breakdowns and the VAT position are books figures
// and must agree with the statements to the rappen.
//
// The two TAX functions are different, deliberately: they are ESTIMATES of a
// position, never postings — a statutory rate times a profit, with a
// coefficient stack (8.5% + 3 1/3% × 232%) that has no exact centime
// representation. They compute in floats like the reference and round to two
// places at the edge, and their outputs go on a snapshot marked with the
// parameters' `confirmed` flags, never into an entry.
//
// One thing the reference could not have: the exercice boundary. Its "YTD"
// filtered on a string year prefix (dev-handoff OPEN-DECISIONS #B2). Here the
// CALLER scopes lines to an exercice before handing them over — the same
// contract as crFor/bilanFor — so a metric cannot mix two years by
// construction.

import { crFor, toCentimes, fromCentimes, type ChartAccount, type CrLine, type PostingLine } from './index'
import type { Money } from '../types'

// ---------------------------------------------------------------------------
// Monthly flows
// ---------------------------------------------------------------------------

/** A posting line that also knows WHEN — flows bucket by month. */
export interface DatedLine extends PostingLine {
  /** The entry's date, `YYYY-MM-DD`. */
  date: string
}

export interface MonthlyFlow {
  /** `YYYY-MM`. */
  month: string
  produits: Money
  charges: Money
}

/**
 * Monthly produits/charges from POSTED lines, CR accounts only. Months with no
 * actual écritures are absent, not zero — the honest series: unposted backlog
 * is invisible here, which is exactly why the recognition worklist matters.
 *
 * Class 3 is produits (credit-nature: credit − debit); every other CR class
 * reads as charges (debit − credit). That is the reference's rule, kept.
 */
export function monthlyFlows(lines: DatedLine[], accounts: ChartAccount[]): MonthlyFlow[] {
  const byNo = new Map(accounts.map((a) => [a.no, a]))
  const m = new Map<string, { produits: bigint; charges: bigint }>()
  for (const l of lines) {
    if (l.status !== 'posted' || l.account_no === null) continue
    const a = byNo.get(l.account_no)
    if (!a || a.statement !== 'cr') continue
    const month = l.date.slice(0, 7)
    const row = m.get(month) ?? { produits: 0n, charges: 0n }
    if (a.class === 3) row.produits += toCentimes(l.credit) - toCentimes(l.debit)
    else row.charges += toCentimes(l.debit) - toCentimes(l.credit)
    m.set(month, row)
  }
  return [...m.keys()]
    .sort()
    .map((k) => ({ month: k, ...m.get(k)! }))
    .filter((r) => r.produits !== 0n || r.charges !== 0n)
    .map((r) => ({ month: r.month, produits: fromCentimes(r.produits), charges: fromCentimes(r.charges) }))
}

/** The simplified book's flows: recettes are produits, dépenses are charges, neutral is neither. */
export function monthlyFlowsRi(rows: { date: string; direction: string; amount: Money }[]): MonthlyFlow[] {
  const m = new Map<string, { produits: bigint; charges: bigint }>()
  for (const r of rows) {
    const month = r.date.slice(0, 7)
    const row = m.get(month) ?? { produits: 0n, charges: 0n }
    if (r.direction === 'recette') row.produits += toCentimes(r.amount)
    else if (r.direction === 'depense') row.charges += toCentimes(r.amount)
    // 'neutral' and anything unknown: in the book, absent from both. index.ts's riTotals rule.
    m.set(month, row)
  }
  return [...m.keys()]
    .sort()
    .map((k) => ({ month: k, ...m.get(k)! }))
    .filter((r) => r.produits !== 0n || r.charges !== 0n)
    .map((r) => ({ month: r.month, produits: fromCentimes(r.produits), charges: fromCentimes(r.charges) }))
}

// ---------------------------------------------------------------------------
// Cost breakdown per management category
// ---------------------------------------------------------------------------

export interface BreakdownInputLine extends DatedLine {
  /** The entry's counterparty, falling back to its raw label — a bar opens into names. */
  counterparty: string
  /** The entry's workspace #number, so a line links back to its écriture. */
  entry_number: number
}

export interface CategoryLine {
  number: number
  date: string
  counterparty: string
  amount: Money
  account: string
}

export interface CategoryBreakdown {
  key: string
  label: unknown
  accounts: string[] | null
  amount: Money
  lines: CategoryLine[]
}

/**
 * Charge breakdown per management category, carrying the underlying ledger
 * lines so every bar can open into its counterparties. POSTED lines only;
 * amount is debit − credit (a credit note reduces its category, honestly).
 * Zero-movement lines are dropped; a category with nothing keeps its row at
 * zero — the set of categories is configuration, not data.
 */
export function costBreakdown(
  categories: { key: string; label: unknown; accounts: string[] }[],
  lines: BreakdownInputLine[]
): CategoryBreakdown[] {
  return categories.map((c) => {
    const out: (CategoryLine & { cents: bigint })[] = []
    for (const l of lines) {
      if (l.status !== 'posted' || l.account_no === null) continue
      if (!c.accounts.includes(l.account_no)) continue
      const cents = toCentimes(l.debit) - toCentimes(l.credit)
      if (cents === 0n) continue
      out.push({
        number: l.entry_number,
        date: l.date,
        counterparty: l.counterparty,
        amount: fromCentimes(cents),
        account: l.account_no,
        cents,
      })
    }
    out.sort((a, b) => (b.cents === a.cents ? 0 : b.cents > a.cents ? 1 : -1))
    const total = out.reduce((s, l) => s + l.cents, 0n)
    return {
      key: c.key,
      label: c.label,
      accounts: c.accounts,
      amount: fromCentimes(total),
      lines: out.map(({ cents: _cents, ...rest }) => rest),
    }
  })
}

/**
 * The simplified book's breakdown: dépenses grouped by the entry's OWN
 * category label (RI entries carry one; accounts do not exist there). An
 * uncategorized dépense lands under a named "Sans catégorie" bucket rather
 * than vanishing — the total must still be the total.
 */
export function costBreakdownRi(
  rows: {
    seq: number
    date: string
    direction: string
    amount: Money
    counterparty: string | null
    raw_label: string
    category: unknown
  }[]
): CategoryBreakdown[] {
  const g = new Map<string, CategoryBreakdown & { cents: bigint }>()
  for (const r of rows) {
    if (r.direction !== 'depense') continue
    const cat = r.category as { fr?: string } | null
    const key = cat?.fr ?? '__none'
    const row =
      g.get(key) ??
      ({
        key,
        label: r.category ?? { fr: 'Sans catégorie', en: 'Uncategorized' },
        accounts: null,
        amount: '0.00',
        lines: [],
        cents: 0n,
      } as CategoryBreakdown & { cents: bigint })
    row.cents += toCentimes(r.amount)
    row.lines.push({
      number: r.seq,
      date: r.date,
      counterparty: r.counterparty ?? r.raw_label,
      amount: r.amount,
      account: '',
    })
    g.set(key, row)
  }
  return [...g.values()]
    .sort((a, b) => (b.cents === a.cents ? 0 : b.cents > a.cents ? 1 : -1))
    .map(({ cents, ...row }) => ({ ...row, amount: fromCentimes(cents) }))
}

// ---------------------------------------------------------------------------
// VAT position — exact, from the entries' TVA columns
// ---------------------------------------------------------------------------

export interface VatEntry {
  status: string
  tva_amount: Money | null
  tva_input_claimed: boolean
  /**
   * This entry's SIGNED movement on revenue accounts, credit-positive: a sale
   * is positive, a credit note (avoir) negative, a pure cost zero.
   *
   * ── IT WAS `credits_revenue: boolean` UNTIL 2026-08-20 ─────────────────────
   * Bala's #65, from sweeping the taxes screen: a credit note declared with
   *
   *   --account 3400 --contra 1020 --tva-rate 8.1
   *
   * stored its 404.63 of VAT correctly and reduced the year's revenue
   * correctly, and the output VAT position ignored it in BOTH directions. A
   * credit note DEBITS revenue, so the old flag was false, and the entry fell
   * through to the input branch — where `tva_input_claimed` is also false, and
   * it was silently dropped. As the report put it: the figure was not wrong by
   * a visible amount, it was wrong by an amount that never appeared.
   *
   * art. 41 al. 1 LTVA settles the bookkeeping question the report deliberately
   * left open: when the consideration is reduced, the taxable person adjusts
   * the output tax. A credit note reduces what the company owes, so the VAT
   * follows the revenue's sign rather than a yes/no about which side it landed
   * on. A boolean could never express that.
   */
  revenue_movement: Money | number
}

export interface VatPosition {
  opening_due: Money
  output_ytd: Money
  input_claimed_ytd: Money
  net_due: Money
}

/**
 * VAT position: output VAT on posted revenue entries, minus input VAT actually
 * CLAIMED (`tva_input_claimed`, the column a CHECK ties to full evidence —
 * art. 26 LTVA), on top of the opening balance of the TVA-due account (2200).
 */
export function vatPosition(opening2200: bigint, entries: VatEntry[]): VatPosition {
  let output = 0n
  let input = 0n
  for (const e of entries) {
    if (e.status !== 'posted') continue
    const tva = toCentimes(e.tva_amount)
    if (tva === 0n) continue
    const revenue = toCentimes(e.revenue_movement)
    // The VAT takes the revenue's sign: a sale adds to what is owed, a credit
    // note subtracts from it (art. 41 LTVA). Only an entry that moves no
    // revenue at all is a candidate for input tax.
    if (revenue > 0n) output += tva
    else if (revenue < 0n) output -= tva
    else if (e.tva_input_claimed) input += tva
  }
  return {
    opening_due: fromCentimes(opening2200),
    output_ytd: fromCentimes(output),
    input_claimed_ytd: fromCentimes(input),
    net_due: fromCentimes(opening2200 + output - input),
  }
}

// ---------------------------------------------------------------------------
// PM tax estimates — floats, on purpose; see the header
// ---------------------------------------------------------------------------

/** The `params` jsonb of books.tax_params. Citations ride along, uninterpreted. */
export interface TaxParams {
  ifd: { rate_pct: number }
  cantonal: { base_rate_pct: number; coefficient_pct: number }
  communal: { coefficient_pct: number }
  capital_tax: { base_rate_permille: number }
}

export interface ProfitTax {
  cantonal: Money
  communal: Money
  ifd: Money
  total: Money
  statutory_pct: number
  effective_pct: number
}

const round2 = (n: number): Money => (Math.round(n * 100) / 100).toFixed(2)

/**
 * PM profit-tax estimate. Statutory combined = IFD + base × (cantonal +
 * communal coefficients); taxes are deductible, so the EFFECTIVE pre-tax rate
 * is s/(1+s). A loss computes as zero, not as a refund.
 */
export function pmProfitTax(profitCentimes: bigint, p: TaxParams): ProfitTax {
  const profit = Math.max(0, Number(profitCentimes) / 100)
  const base = profit * (p.cantonal.base_rate_pct / 100)
  const cantonal = base * (p.cantonal.coefficient_pct / 100)
  const communal = base * (p.communal.coefficient_pct / 100)
  const ifd = profit * (p.ifd.rate_pct / 100)
  const statutoryPct =
    p.ifd.rate_pct + p.cantonal.base_rate_pct * ((p.cantonal.coefficient_pct + p.communal.coefficient_pct) / 100)
  const effectivePct = statutoryPct / (1 + statutoryPct / 100)
  return {
    cantonal: round2(cantonal),
    communal: round2(communal),
    ifd: round2(ifd),
    total: round2(cantonal + communal + ifd),
    statutory_pct: Math.round(statutoryPct * 100) / 100,
    effective_pct: Math.round(effectivePct * 100) / 100,
  }
}

export interface CapitalTax {
  gross: Money
  credited: Money
  net_due: Money
}

/**
 * Capital-tax estimate with the art. 118 LI-VD imputation shown, not hidden:
 * `gross` is the 0.6‰ of book equity, `credited` is the cantonal+communal
 * profit tax counted against it, `net_due` is what remains — it only bites in
 * loss or low-profit years. Whether the imputation applies exactly this way is
 * the parameters' OPEN QUESTION (capital_tax.confirmed stays false until the
 * fiduciary answers — decided with Mustneer 2026-08-19); serving all three
 * figures is what lets a reader apply either reading.
 */
export function pmCapitalTax(equityCentimes: bigint, cantCommProfitTax: number, p: TaxParams): CapitalTax {
  const equity = Math.max(0, Number(equityCentimes) / 100)
  const gross = equity * (p.capital_tax.base_rate_permille / 1000)
  const credited = Math.min(gross, cantCommProfitTax)
  return { gross: round2(gross), credited: round2(credited), net_due: round2(Math.max(0, gross - credited)) }
}

// ---------------------------------------------------------------------------
// The compte de résultat, month by month
// ---------------------------------------------------------------------------

/**
 * One month's compte de résultat, in the statutory line structure.
 *
 * ===========================================================================
 * A MONTHLY P&L IS A READING AID, NOT A STATEMENT
 * ===========================================================================
 * art. 959b fixes the ANNUAL compte de résultat. A month is not a legal
 * reporting period: nothing closes at a month boundary, no result is carried at
 * one, and no month of this is filable. It lives here, beside the other
 * management derivations, rather than in `derive/index.ts` with the statutory
 * statements — and that placement is the documentation.
 *
 * What it is for: the operator asked to see where a year went (ticket #64). The
 * annual statement answers "the year lost 10'993.60" and cannot answer "and
 * almost all of it was March".
 */
export interface MonthlyCr {
  /** `YYYY-MM`. */
  month: string
  lines: CrLine[]
  resultat: Money
}

/**
 * Split the year into months and derive the real statement for each.
 *
 * ── EVERY MONTH IN THE SPAN, INCLUDING THE EMPTY ONES ──────────────────────
 * `monthlyFlows` drops months with no movement, which is right for a sparkline
 * and wrong for a grid: a table whose columns appear and disappear cannot be
 * read across, and the reader cannot tell "no trading" from "no data". So the
 * caller passes the exercice's own boundaries and gets one column per month
 * between them, a quiet month coming back as a full set of zero lines.
 *
 * That is the same rule the annual statement already follows — "every legal
 * line is emitted, including the ones that come to zero" — applied to the
 * second axis.
 *
 * ── THE STRUCTURE IS DERIVED, NEVER SPLIT ──────────────────────────────────
 * Each month runs through `crFor`, the same function the annual statement uses.
 * Nothing here knows what a statutory line is, so a change to art. 959b's
 * structure reaches the monthly view for free, and the two can never disagree
 * about what a line means. Summing the twelve months gives the annual result
 * exactly, because it is the same arithmetic over a partition of the same rows.
 */
export function crByMonth(
  lines: DatedLine[],
  accounts: ChartAccount[],
  span: { starts_on: string; ends_on: string }
): MonthlyCr[] {
  const byMonth = new Map<string, DatedLine[]>()
  for (const m of monthsBetween(span.starts_on, span.ends_on)) byMonth.set(m, [])
  for (const l of lines) {
    const month = l.date.slice(0, 7)
    // A line dated outside the exercice cannot happen (the entry door refuses
    // it), but a derivation that silently dropped one would hide the day it
    // does. It gets its own column instead.
    const bucket = byMonth.get(month)
    if (bucket) bucket.push(l)
    else byMonth.set(month, [l])
  }
  return [...byMonth.keys()].sort().map((month) => {
    const cr = crFor(byMonth.get(month)!, accounts)
    return { month, lines: cr.lines, resultat: cr.resultat }
  })
}

/** `2026-01` … `2026-12` for a calendar year; whatever the span really is otherwise. */
function monthsBetween(startsOn: string, endsOn: string): string[] {
  const out: string[] = []
  let y = Number(startsOn.slice(0, 4))
  let m = Number(startsOn.slice(5, 7))
  const last = endsOn.slice(0, 7)
  for (let guard = 0; guard < 240; guard++) {
    const cur = `${y}-${String(m).padStart(2, '0')}`
    out.push(cur)
    if (cur >= last) break
    m += 1
    if (m > 12) {
      m = 1
      y += 1
    }
  }
  return out
}
